File: 0069-sqrtx/0069-sqrtx.c
int mySqrt(int x) {
    if (x == 0 || x == 1)
        return x;

    int left = 1, right = x, ans = 0;

    while (left <= right) {
        int mid = left + (right - left) / 2;

        if (mid <= x / mid) {
            ans = mid;        // possible answer
            left = mid + 1;   // try bigger number
        } else {
            right = mid - 1;  // try smaller number
        }
    }

    return ans;
}

File: 0069-sqrtx/0069-sqrtx.java
// class Solution {
//     public int mySqrt(int x) {
//         if (x == 0 || x == 1)
//             return x;
//         int low = 1;
//         int high = x;
//         int ans = 1;

//         while(low <= high){
//             int mid = low + (high - low) / 2;
//             if(mid * mid > x){
//                 high = mid - 1;
//             }
//             else{
//                 low = mid + 1;
//                 ans = mid;
//             }
//         }
//         return ans;
//     }
// }

class Solution {
    public int mySqrt(int x) {
        if (x == 0 || x == 1)
            return x;

        int low = 1;
        int high = x;
        int ans = 1;

        while (low <= high) {
            int mid = low + (high - low) / 2;

            if ((long) mid * mid > x) {
                high = mid - 1;
            } 
            else {
                ans = mid;
                low = mid + 1;
            }
        }

        return ans;
    }
}